**VIDEO LINK ::::::>>>>**

https://drive.google.com/file/d/1zYaVdtgnABK8oScmJ2e2fWdNtaE83siS/view?usp=sharing

In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [ ]:
!pip install attrdict
!pip install deepdish

In [ ]:
!pip install http://download.pytorch.org/whl/cpu/torch-0.4.0-cp27-cp27mu-linux_x86_64.whl 
!pip install pyro-ppl

ERROR: torch-0.4.0-cp27-cp27mu-linux_x86_64.whl is not a supported wheel on this platform.
     |████████████████████████████████| 718 kB 4.7 MB/s 
     |████████████████████████████████| 750.6 MB 12 kB/s 
  Attempting uninstall: torch
    Found existing installation: torch 1.10.0+cu111
    Uninstalling torch-1.10.0+cu111:
      Successfully uninstalled torch-1.10.0+cu111
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchvision 0.11.1+cu111 requires torch==1.10.0, but you have torch 1.11.0 which is incompatible.
torchtext 0.11.0 requires torch==1.10.0, but you have torch 1.11.0 which is incompatible.
torchaudio 0.10.0+cu111 requires torch==1.10.0, but you have torch 1.11.0 which is incompatible.


In [ ]:
!pip install uncertainty-calibration


  Created wheel for uncertainty-calibration: filename=uncertainty_calibration-0.0.9-py3-none-any.whl size=12286 sha256=8761e111669bb7485556e502fa918db510d700b4fb269c89000e1fcd69a4e5ed
  Stored in directory: /root/.cache/pip/wheels/d2/b4/25/3eee8271e3c0d7b2d1f3e4e3312daf1c981405a8526860eb09
Successfully built uncertainty-calibration


In [ ]:
!python '/content/drive/MyDrive/conf_matrix_and_calibration-main/experiments/calibrate_combo_experiment.py'

Models:   0% 0/4 [00:00<?, ?it/s]
Runs:   0% 0/25 [00:00<?, ?it/s]
Runs:   4% 1/25 [00:05<02:05,  5.25s/it]
Runs:   8% 2/25 [00:10<02:01,  5.27s/it]
Runs:  12% 3/25 [00:14<01:41,  4.63s/it]
Runs:  16% 4/25 [00:16<01:16,  3.65s/it]
Runs:  20% 5/25 [00:18<01:03,  3.18s/it]
Runs:  24% 6/25 [00:21<00:54,  2.85s/it]
Runs:  28% 7/25 [00:22<00:45,  2.52s/it]
Runs:  32% 8/25 [00:25<00:40,  2.39s/it]
Runs:  36% 9/25 [00:27<00:37,  2.35s/it]
Runs:  40% 10/25 [00:29<00:34,  2.33s/it]
Runs:  44% 11/25 [00:31<00:30,  2.16s/it]
Runs:  48% 12/25 [00:33<00:29,  2.23s/it]
Runs:  52% 13/25 [00:36<00:27,  2.26s/it]
Runs:  56% 14/25 [00:38<00:24,  2.22s/it]
Runs:  60% 15/25 [00:40<00:22,  2.22s/it]
Runs:  64% 16/25 [00:42<00:19,  2.21s/it]
Runs:  68% 17/25 [00:44<00:17,  2.13s/it]
Runs:  72% 18/25 [00:46<00:15,  2.17s/it]
Runs:  76% 19/25 [00:48<00:12,  2.14s/it]
Runs:  80% 20/25 [00:50<00:10,  2.10s/it]
Runs:  84% 21/25 [00:52<00:08,  2.08s/it]
Runs:  88% 22/25 [00:55<00:06,  2.13s/it]
Runs:  92% 23/25 [

In [ ]:

!python '/content/drive/MyDrive/conf_matrix_and_calibration-main/experiments/calibration_experiment.py'

Models:   0% 0/4 [00:00<?, ?it/s]
Runs:   0% 0/25 [00:00<?, ?it/s]
Runs:   4% 1/25 [00:04<01:36,  4.04s/it]
Runs:   8% 2/25 [00:07<01:29,  3.90s/it]
Runs:  12% 3/25 [00:11<01:25,  3.89s/it]
Runs:  16% 4/25 [00:15<01:21,  3.86s/it]
Runs:  20% 5/25 [00:19<01:16,  3.80s/it]
Runs:  24% 6/25 [00:22<01:11,  3.76s/it]
Runs:  28% 7/25 [00:26<01:08,  3.80s/it]
Runs:  32% 8/25 [00:30<01:04,  3.82s/it]
Runs:  36% 9/25 [00:34<01:00,  3.81s/it]
Runs:  40% 10/25 [00:38<00:57,  3.81s/it]
Runs:  44% 11/25 [00:42<00:53,  3.79s/it]
Runs:  48% 12/25 [00:45<00:49,  3.78s/it]
Runs:  52% 13/25 [00:49<00:45,  3.80s/it]
Runs:  56% 14/25 [00:53<00:41,  3.81s/it]
Runs:  60% 15/25 [00:57<00:38,  3.83s/it]
Runs:  64% 16/25 [01:01<00:34,  3.80s/it]
Runs:  68% 17/25 [01:04<00:30,  3.80s/it]
Runs:  72% 18/25 [01:08<00:26,  3.82s/it]
Runs:  76% 19/25 [01:12<00:22,  3.79s/it]
Runs:  80% 20/25 [01:16<00:18,  3.77s/it]
Runs:  84% 21/25 [01:19<00:14,  3.74s/it]
Runs:  88% 22/25 [01:23<00:11,  3.73s/it]
Runs:  92% 23/25 [

###Q.1. Reproduce Results: [25 points]
Reproduce the results depicted in the paper only for CIFAR-10 dataset using the authors’ code.
Out of the four pre-trained CNN models considered in the paper for CIFAR-10 experiments, it will
okay if you consider any two dissimilar approaches.
Are you able to get the same results as reported by the authors? Comment on the challenges
faced. List all the hyper-parameters values that authors didn’t mention and you need to make your
own choice for the same.

### Results of calibration_experiment

In [ ]:
from google.colab import files
 
 
uploaded = files.upload()

Saving experimentsdensenet-bc-L190-k40_accuracy.csv to experimentsdensenet-bc-L190-k40_accuracy.csv


In [ ]:
import pandas as pd
import io

calibdensenet_acc = pd.read_csv(io.BytesIO(uploaded['experimentsdensenet-bc-L190-k40_accuracy.csv']))
(calibdensenet_acc)

human     model  comb none  comb confusion  comb temperature scaling  \
0   0.950000  0.963667   0.973667        0.962000                  0.978000   
1   0.950333  0.966333   0.977333        0.964333                  0.977333   
2   0.953667  0.970333   0.978667        0.967333                  0.980333   
3   0.954667  0.972333   0.981333        0.971333                  0.982000   
4   0.955333  0.964333   0.977000        0.958333                  0.979333   
5   0.954667  0.968667   0.978667        0.967333                  0.981000   
6   0.956000  0.969333   0.979667        0.970333                  0.979333   
7   0.953333  0.971000   0.978333        0.971000                  0.979667   
8   0.952000  0.965333   0.975333        0.964667                  0.979000   
9   0.956000  0.966333   0.978333        0.966667                  0.978667   
10  0.953000  0.967333   0.976000        0.964667                  0.978667   
11  0.955333  0.972333   0.980000        0.968667                  0.981000   
12  0.947333  0.964000   0.973333        0.965000                  0.976000   
13  0.951667  0.965667   0.978333        0.968333                  0.978667   
14  0.958667  0.965000   0.979333        0.967333                  0.981667   
15  0.958333  0.968667   0.979333        0.966000                  0.981667   
16  0.952333  0.968333   0.978333        0.967667                  0.981000   
17  0.949667  0.963000   0.975333        0.962000                  0.976667   
18  0.958333  0.967000   0.976000        0.966000                  0.979667   
19  0.951333  0.969000   0.979000        0.965000                  0.980667   
20  0.950333  0.968333   0.975667        0.965000                  0.977333   
21  0.953667  0.973333   0.981333        0.967667                  0.981000   
22  0.952667  0.967667   0.979667        0.964000                  0.980000   
23  0.955667  0.962000   0.977667        0.961667                  0.981000   
24  0.961333  0.967667   0.978000        0.970667                  0.982333   

    comb ensemble temperature scaling  comb imax binning  
0                            0.977000           0.977667  
1                            0.977333           0.977333  
2                            0.979333           0.979667  
3                            0.979000           0.982333  
4                            0.979000           0.979667  
5                            0.980333           0.982000  
6                            0.977667           0.979000  
7                            0.978333           0.978333  
8                            0.976667           0.979000  
9                            0.977333           0.978667  
10                           0.977000           0.978000  
11                           0.979667           0.982333  
12                           0.976667           0.976333  
13                           0.979000           0.980333  
14                           0.978333           0.980000  
15                           0.980000           0.979333  
16                           0.979000           0.980333  
17                           0.974000           0.976000  
18                           0.979000           0.979333  
19                           0.979667           0.980000  
20                           0.976000           0.977667  
21                           0.978333           0.980667  
22                           0.977667           0.980667  
23                           0.980000           0.981333  
24                           0.982333           0.983000

In [ ]:
uploaded = files.upload()

Saving experimentsdensenet-bc-L190-k40_calibration.csv to experimentsdensenet-bc-L190-k40_calibration.csv


In [ ]:
calibdensenet_calib = pd.read_csv(io.BytesIO(uploaded['experimentsdensenet-bc-L190-k40_calibration.csv']))
(calibdensenet_calib)

calibration method       metric     model      comb
0                 none      ECE (W)  0.028971  0.020763
1                 none      ECE (M)  0.026739  0.020356
2                 none   cwECE (WT)  0.040232  0.028296
3                 none   cwECE (MT)  0.024069  0.018143
4                 none  cwECE (WNT)  0.006635  0.004745
..                 ...          ...       ...       ...
870       imax binning   cwECE (WT)  0.025019  0.020849
871       imax binning   cwECE (MT)  0.014573  0.008784
872       imax binning  cwECE (WNT)  0.003570  0.002900
873       imax binning  cwECE (MNT)  0.001939  0.001131
874       imax binning          NLL  0.117004  0.353385

[875 rows x 4 columns]

In [ ]:
uploaded = files.upload()

Saving experimentspreresnet-110_accuracy.csv to experimentspreresnet-110_accuracy.csv


In [ ]:
calibpreresnet_acc = pd.read_csv(io.BytesIO(uploaded['experimentspreresnet-110_accuracy.csv']))
(calibpreresnet_acc)

human     model  comb none  comb confusion  comb temperature scaling  \
0   0.950000  0.948667   0.968667        0.954667                  0.971000   
1   0.950333  0.950667   0.969667        0.954667                  0.976000   
2   0.953667  0.949000   0.972000        0.950667                  0.974000   
3   0.954667  0.953000   0.972667        0.961333                  0.978667   
4   0.955333  0.944000   0.966667        0.955667                  0.974667   
5   0.954667  0.952667   0.972667        0.960667                  0.976000   
6   0.956000  0.956000   0.974333        0.962333                  0.978333   
7   0.953333  0.950667   0.973333        0.957000                  0.977667   
8   0.952000  0.951000   0.971333        0.956333                  0.974667   
9   0.956000  0.950000   0.970000        0.959667                  0.974667   
10  0.953000  0.952667   0.970667        0.957000                  0.973333   
11  0.955333  0.958333   0.974000        0.962000                  0.979000   
12  0.947333  0.949333   0.969667        0.950333                  0.974333   
13  0.951667  0.948333   0.974000        0.957667                  0.975333   
14  0.958667  0.955333   0.972667        0.962333                  0.974667   
15  0.958333  0.952667   0.969667        0.960333                  0.976333   
16  0.952333  0.946667   0.970000        0.957000                  0.977000   
17  0.949667  0.949667   0.969667        0.951000                  0.973667   
18  0.958333  0.947000   0.970333        0.955000                  0.976667   
19  0.951333  0.944333   0.967667        0.952667                  0.972667   
20  0.950333  0.947667   0.968667        0.952000                  0.970667   
21  0.953667  0.952333   0.973667        0.955000                  0.976000   
22  0.952667  0.946000   0.968667        0.952333                  0.974000   
23  0.955667  0.944667   0.970667        0.951667                  0.976667   
24  0.961333  0.950000   0.973667        0.960333                  0.981000   

    comb ensemble temperature scaling  comb imax binning  
0                            0.971333           0.970333  
1                            0.976333           0.975000  
2                            0.974000           0.974667  
3                            0.978667           0.977667  
4                            0.974667           0.974000  
5                            0.976333           0.977000  
6                            0.978333           0.978000  
7                            0.977667           0.977333  
8                            0.974667           0.974333  
9                            0.973667           0.974000  
10                           0.973000           0.973667  
11                           0.979000           0.979000  
12                           0.974667           0.974667  
13                           0.975000           0.977000  
14                           0.974333           0.974667  
15                           0.976333           0.977000  
16                           0.977333           0.977333  
17                           0.973333           0.973000  
18                           0.976000           0.977333  
19                           0.972333           0.972667  
20                           0.970333           0.971333  
21                           0.976000           0.979333  
22                           0.974333           0.974000  
23                           0.977000           0.978333  
24                           0.981333           0.981333

In [ ]:
uploaded = files.upload()

Saving experimentspreresnet-110_calibration.csv to experimentspreresnet-110_calibration.csv


In [ ]:
calibpreresnet_calib = pd.read_csv(io.BytesIO(uploaded['experimentspreresnet-110_calibration.csv']))
(calibpreresnet_calib)

calibration method       metric     model      comb
0                 none      ECE (W)  0.030581  0.021799
1                 none      ECE (M)  0.030187  0.020010
2                 none   cwECE (WT)  0.047635  0.031994
3                 none   cwECE (MT)  0.026016  0.015749
4                 none  cwECE (WNT)  0.007819  0.004995
..                 ...          ...       ...       ...
870       imax binning   cwECE (WT)  0.030479  0.019738
871       imax binning   cwECE (MT)  0.016856  0.009189
872       imax binning  cwECE (WNT)  0.004583  0.002902
873       imax binning  cwECE (MNT)  0.002756  0.001204
874       imax binning          NLL  0.163614  0.365280

[875 rows x 4 columns]

In [ ]:
#Calculate error
import numpy as np
model_calib = ['PreResNet-164', 'DenseNet-BC']
model_res = [(1 - np.mean(calibpreresnet_acc['model']))*100, (1 - np.mean(calibdensenet_acc['model']))*100]
print(model_res)
TS_calib = [(1 - np.mean(calibpreresnet_acc['comb temperature scaling']))*100, (1 - np.mean(calibdensenet_acc['comb temperature scaling']))*100]
print(TS_calib)
ETS_calib = [(1 - np.mean(calibpreresnet_acc['comb ensemble temperature scaling']))*100, (1 - np.mean(calibdensenet_acc['comb ensemble temperature scaling']))*100]
print(ETS_calib)
IMax_calib = [(1 - np.mean(calibpreresnet_acc['comb imax binning']))*100, (1 - np.mean(calibdensenet_acc['comb imax binning']))*100]
print(IMax_calib)

[4.997333333333332, 3.2519999999999993]
[2.4519999999999986, 2.031999999999967]
[2.4560000000000137, 2.1653333333333413]
[2.428000000000008, 2.0440000000000014]


In [ ]:
from tabulate import tabulate
info = {'Model Name': model_calib, 'Model': model_res, 'TS': TS_calib, 'ETS': ETS_calib, 'IMax': IMax_calib}
print(tabulate(info, headers='keys'))

Model Name       Model     TS      ETS    IMax
-------------  -------  -----  -------  ------
PreResNet-164  4.99733  2.452  2.456     2.428
DenseNet-BC    3.252    2.032  2.16533   2.044


**OBSERVATIONS:**
* If we compare the results from above table with Appendix C of paper, we can see that we get similar results for PreResNet-164 & DenseNet-BC.
* For PreResNet-164 in Paper: 
  * Model = 5.00 ± 0.36
  * TS = 2.43 ± 0.22  
  * ETS = 2.46 ± 0.25
  * IMax = 2.43 ± 0.26

* For DenseNet-BC in Paper:
  * Model = 3.25 ± 0.30
  * TS =  2.01 ± 0.15
  * ETS =  2.17 ± 0.17
  * IMax =  2.04 ± 0.18
* So, the results given in paper are very similar to the results we got in the above table





###Results of calibrate_combo_experiment

In [ ]:
from google.colab import files
 
 
uploaded = files.upload()

Saving densenet-bc-L190-k40_accuracy.csv to densenet-bc-L190-k40_accuracy.csv


In [ ]:
import pandas as pd
import io

 
densenet_combo= pd.read_csv(io.BytesIO(uploaded['densenet-bc-L190-k40_accuracy.csv']))
(densenet_combo)

human     model  comb none  comb confusion  comb temperature scaling
0   0.950000  0.963667   0.977667        0.975667                       NaN
1   0.950000  0.963667   0.977667        0.975667                       NaN
2   0.950333  0.966333   0.979333        0.978667                       NaN
3   0.950333  0.966333   0.979333        0.978667                       NaN
4   0.953667  0.970333   0.981667        0.982333                       NaN
5   0.953667  0.970333   0.981667        0.982333                       NaN
6   0.954667  0.972333   0.981333        0.981333                       NaN
7   0.954667  0.972333   0.981333        0.981333                       NaN
8   0.955333  0.964333   0.979667        0.977667                       NaN
9   0.955333  0.964333   0.979667        0.977667                       NaN
10  0.954667  0.968667   0.979000        0.979667                       NaN
11  0.954667  0.968667   0.979000        0.979667                       NaN
12  0.956000  0.969333   0.982333        0.981333                       NaN
13  0.956000  0.969333   0.982333        0.981333                       NaN
14  0.953333  0.971000   0.980000        0.978667                       NaN
15  0.953333  0.971000   0.980000        0.978667                       NaN
16  0.952000  0.965333   0.978333        0.978333                       NaN
17  0.952000  0.965333   0.978333        0.978333                       NaN
18  0.956000  0.966333   0.979333        0.979667                       NaN
19  0.956000  0.966333   0.979333        0.979667                       NaN
20  0.953000  0.967333   0.981000        0.978667                       NaN
21  0.953000  0.967333   0.981000        0.978667                       NaN
22  0.955333  0.972333   0.982000        0.981333                       NaN
23  0.955333  0.972333   0.982000        0.981333                       NaN
24  0.947333  0.964000   0.974667        0.974333                       NaN
25  0.947333  0.964000   0.974667        0.974333                       NaN
26  0.951667  0.965667   0.979667        0.980000                       NaN
27  0.951667  0.965667   0.979667        0.980000                       NaN
28  0.958667  0.965000   0.981333        0.980333                       NaN
29  0.958667  0.965000   0.981333        0.980333                       NaN
30  0.958333  0.968667   0.982333        0.982333                       NaN
31  0.958333  0.968667   0.982333        0.982333                       NaN
32  0.952333  0.968333   0.979667        0.979333                       NaN
33  0.952333  0.968333   0.979667        0.979333                       NaN
34  0.949667  0.963000   0.979333        0.976333                       NaN
35  0.949667  0.963000   0.979333        0.976333                       NaN
36  0.958333  0.967000   0.979333        0.977667                       NaN
37  0.958333  0.967000   0.979333        0.977667                       NaN
38  0.951333  0.969000   0.979667        0.979667                       NaN
39  0.951333  0.969000   0.979667        0.979667                       NaN
40  0.950333  0.968333   0.976000        0.975667                       NaN
41  0.950333  0.968333   0.976000        0.975667                       NaN
42  0.953667  0.973333   0.982667        0.983000                       NaN
43  0.953667  0.973333   0.982667        0.983000                       NaN
44  0.952667  0.967667   0.982667        0.981000                       NaN
45  0.952667  0.967667   0.982667        0.981000                       NaN
46  0.955667  0.962000   0.981667        0.978667                       NaN
47  0.955667  0.962000   0.981667        0.978667                       NaN
48  0.961333  0.967667   0.981333        0.980000                       NaN
49  0.961333  0.967667   0.981333        0.980000                       NaN

In [ ]:

 
 
uploaded = files.upload()

Saving densenet-bc-L190-k40_calibration.csv to densenet-bc-L190-k40_calibration.csv


In [ ]:

 
densenet_combo2 = pd.read_csv(io.BytesIO(uploaded['densenet-bc-L190-k40_calibration.csv']))
(densenet_combo2 )

calibration method       metric     model      comb  comb (post cal)
0               MAP_CI      ECE (W)  0.022666  0.015448         0.011258
1               MAP_CI      ECE (M)  0.021948  0.014656         0.011407
2               MAP_CI   cwECE (WT)  0.037662  0.026239         0.024011
3               MAP_CI   cwECE (MT)  0.018976  0.011796         0.010343
4               MAP_CI  cwECE (WNT)  0.005765  0.003802         0.003275
..                 ...          ...       ...       ...              ...
520       uncal_MAP_CI   cwECE (WT)  0.037264  0.021687         0.023955
521       uncal_MAP_CI   cwECE (MT)  0.020885  0.013537         0.010501
522       uncal_MAP_CI  cwECE (WNT)  0.005886  0.003541         0.003387
523       uncal_MAP_CI  cwECE (MNT)  0.001696  0.000862         0.000862
524       uncal_MAP_CI          NLL  0.161318  0.082792         0.067616

[525 rows x 5 columns]

In [ ]:

 
 
uploaded = files.upload()

Saving preresnet-110_accuracy.csv to preresnet-110_accuracy.csv


In [ ]:

 
preresnet_combo = pd.read_csv(io.BytesIO(uploaded['preresnet-110_accuracy.csv']))
(preresnet_combo)

human     model  comb none  comb confusion  comb temperature scaling
0   0.950000  0.948667   0.973000        0.970667                       NaN
1   0.950000  0.948667   0.973000        0.970667                       NaN
2   0.950333  0.950667   0.977000        0.972333                       NaN
3   0.950333  0.950667   0.977000        0.972333                       NaN
4   0.953667  0.949000   0.977000        0.973333                       NaN
5   0.953667  0.949000   0.977000        0.973333                       NaN
6   0.954667  0.953000   0.977000        0.973000                       NaN
7   0.954667  0.953000   0.977000        0.973000                       NaN
8   0.955333  0.944000   0.971667        0.967333                       NaN
9   0.955333  0.944000   0.971667        0.967333                       NaN
10  0.954667  0.952667   0.976000        0.974667                       NaN
11  0.954667  0.952667   0.976000        0.974667                       NaN
12  0.956000  0.956000   0.978667        0.976667                       NaN
13  0.956000  0.956000   0.978667        0.976667                       NaN
14  0.953333  0.950667   0.977667        0.974000                       NaN
15  0.953333  0.950667   0.977667        0.974000                       NaN
16  0.952000  0.951000   0.975667        0.973667                       NaN
17  0.952000  0.951000   0.975667        0.973667                       NaN
18  0.956000  0.950000   0.973667        0.971000                       NaN
19  0.956000  0.950000   0.973667        0.971000                       NaN
20  0.953000  0.952667   0.975333        0.973000                       NaN
21  0.953000  0.952667   0.975333        0.973000                       NaN
22  0.955333  0.958333   0.978000        0.975667                       NaN
23  0.955333  0.958333   0.978000        0.975667                       NaN
24  0.947333  0.949333   0.971333        0.970000                       NaN
25  0.947333  0.949333   0.971333        0.970000                       NaN
26  0.951667  0.948333   0.976667        0.974000                       NaN
27  0.951667  0.948333   0.976667        0.974000                       NaN
28  0.958667  0.955333   0.976000        0.974000                       NaN
29  0.958667  0.955333   0.976000        0.974000                       NaN
30  0.958333  0.952667   0.978667        0.972667                       NaN
31  0.958333  0.952667   0.978667        0.972667                       NaN
32  0.952333  0.946667   0.975667        0.971000                       NaN
33  0.952333  0.946667   0.975667        0.971000                       NaN
34  0.949667  0.949667   0.975333        0.972667                       NaN
35  0.949667  0.949667   0.975333        0.972667                       NaN
36  0.958333  0.947000   0.973667        0.971333                       NaN
37  0.958333  0.947000   0.973667        0.971333                       NaN
38  0.951333  0.944333   0.973667        0.969000                       NaN
39  0.951333  0.944333   0.973667        0.969000                       NaN
40  0.950333  0.947667   0.971000        0.969667                       NaN
41  0.950333  0.947667   0.971000        0.969667                       NaN
42  0.953667  0.952333   0.974667        0.973667                       NaN
43  0.953667  0.952333   0.974667        0.973667                       NaN
44  0.952667  0.946000   0.974333        0.969667                       NaN
45  0.952667  0.946000   0.974333        0.969667                       NaN
46  0.955667  0.944667   0.976667        0.972333                       NaN
47  0.955667  0.944667   0.976667        0.972333                       NaN
48  0.961333  0.950000   0.978333        0.975333                       NaN
49  0.961333  0.950000   0.978333        0.975333                       NaN

In [ ]:

 
uploaded = files.upload()

Saving preresnet-110_calibration.csv to preresnet-110_calibration.csv


In [ ]:
preresnet_combo2 = pd.read_csv(io.BytesIO(uploaded['preresnet-110_calibration.csv']))
(preresnet_combo2)

calibration method       metric     model      comb  comb (post cal)
0               MAP_CI      ECE (W)  0.023111  0.016885         0.012632
1               MAP_CI      ECE (M)  0.022079  0.016223         0.012304
2               MAP_CI   cwECE (WT)  0.042405  0.028291         0.026480
3               MAP_CI   cwECE (MT)  0.019980  0.012375         0.009195
4               MAP_CI  cwECE (WNT)  0.006816  0.004205         0.003745
..                 ...          ...       ...       ...              ...
520       uncal_MAP_CI   cwECE (WT)  0.046754  0.028928         0.026816
521       uncal_MAP_CI   cwECE (MT)  0.028501  0.011175         0.010601
522       uncal_MAP_CI  cwECE (WNT)  0.007904  0.003956         0.003488
523       uncal_MAP_CI  cwECE (MNT)  0.002901  0.001220         0.001179
524       uncal_MAP_CI          NLL  0.213760  0.093139         0.077943

[525 rows x 5 columns]

In [ ]:
# Use groupby to comine and show results in a table
metric_pre = preresnet_combo2.groupby('metric')
metric_dense = densenet_combo2.groupby('metric')

In [ ]:
metric_ECEW_p = metric_pre.get_group('ECE (W)')
metric_ECEM_p = metric_pre.get_group('ECE (M)')
metric_cwECEWT_p = metric_pre.get_group('cwECE (WT)')
metric_cwECEMT_p = metric_pre.get_group('cwECE (MT)')
metric_cwECEWNT_p = metric_pre.get_group('cwECE (WNT)')
metric_cwECEMNT_p = metric_pre.get_group('cwECE (MNT)')
metric_NLL_p = metric_pre.get_group('NLL')

metric_ECEW_d = metric_dense.get_group('ECE (W)')
metric_ECEM_d = metric_dense.get_group('ECE (M)')
metric_cwECEWT_d = metric_dense.get_group('cwECE (WT)')
metric_cwECEMT_d = metric_dense.get_group('cwECE (MT)')
metric_cwECEWNT_d = metric_dense.get_group('cwECE (WNT)')
metric_cwECEMNT_d = metric_dense.get_group('cwECE (MNT)')
metric_NLL_d = metric_dense.get_group('NLL')


In [ ]:
#Caluclte the mean error
import numpy as np
model_ece = [1/2*(np.mean(metric_ECEW_p['model']) + np.mean(metric_ECEM_p['model']))*100 ,1/2*(np.mean(metric_ECEW_d['model']) + np.mean(metric_ECEM_d['model']))*100]
model_cwece = [1/4*(np.mean(metric_cwECEWT_p['model']) + np.mean(metric_cwECEMT_p['model']) + np.mean(metric_cwECEWNT_p['model'])+ np.mean(metric_cwECEMNT_p['model']))*100 , 1/4*(np.mean(metric_cwECEWT_d['model']) + np.mean(metric_cwECEMT_d['model']) + np.mean(metric_cwECEWNT_d['model']) + np.mean(metric_cwECEMNT_d['model']))*100]   
model_NLL = [np.mean(metric_NLL_p['model']), np.mean(metric_NLL_d['model']) ]        

comb_ece = [1/2*(np.mean(metric_ECEW_p['comb']) + np.mean(metric_ECEM_p['comb']))*100 ,1/2*(np.mean(metric_ECEW_d['comb']) + np.mean(metric_ECEM_d['comb']))*100]
comb_cwece = [1/4*(np.mean(metric_cwECEWT_p['comb']) + np.mean(metric_cwECEMT_p['comb']) + np.mean(metric_cwECEWNT_p['comb'])+ np.mean(metric_cwECEMNT_p['comb']))*100 , 1/4*(np.mean(metric_cwECEWT_d['comb']) + np.mean(metric_cwECEMT_d['comb']) + np.mean(metric_cwECEWNT_d['comb'])+ np.mean(metric_cwECEMNT_d['comb']))*100]   
comb_NLL = [np.mean(metric_NLL_p['comb']), np.mean(metric_NLL_d['comb'])] 

recal_ece = [1/2*(np.mean(metric_ECEW_p['comb (post cal)']) + np.mean(metric_ECEM_p['comb (post cal)']))*100 , 1/2*(np.mean(metric_ECEW_d['comb (post cal)']) + np.mean(metric_ECEM_d['comb (post cal)']))*100]
recal_cwece = [1/4*(np.mean(metric_cwECEWT_p['comb (post cal)']) + np.mean(metric_cwECEMT_p['comb (post cal)']) + np.mean(metric_cwECEWNT_p['comb (post cal)'])+ np.mean(metric_cwECEMNT_p['comb (post cal)']))*100 , 1/4*(np.mean(metric_cwECEWT_d['comb (post cal)']) + np.mean(metric_cwECEMT_d['comb (post cal)']) + np.mean(metric_cwECEWNT_d['comb (post cal)'])+ np.mean(metric_cwECEMNT_d['comb (post cal)']))*100]   
recal_NLL = [np.mean(metric_NLL_p['comb (post cal)']), np.mean(metric_NLL_d['comb (post cal)']) ] 

In [ ]:
from tabulate import tabulate
metric_name = ['ECE', 'cwECE', 'NLL']
model_calib = ['PreResNet-164', 'DenseNet-BC']
print('ECE :')
info = {'Model Name': model_calib, 'Model':model_ece, 'Comb.': comb_ece, 'Recal. Comb': recal_ece}
print(tabulate(info, headers='keys'))
print('\n\ncwECE :')
info = {'Model Name': model_calib, 'Model':model_cwece, 'Comb.': comb_cwece, 'Recal. Comb': recal_cwece}
print(tabulate(info, headers='keys'))
print('\n\nNLL :')
info = {'Model Name': model_calib, 'Model':model_NLL, 'Comb.': comb_NLL, 'Recal. Comb': recal_NLL}
print(tabulate(info, headers='keys'))

ECE :
Model Name       Model    Comb.    Recal. Comb
-------------  -------  -------  -------------
PreResNet-164  2.56874  1.62835        1.29225
DenseNet-BC    1.96987  1.38542        1.11889


cwECE :
Model Name        Model     Comb.    Recal. Comb
-------------  --------  --------  -------------
PreResNet-164  0.190336  0.118874      0.110564
DenseNet-BC    0.151954  0.10011       0.0954192


NLL :
Model Name        Model      Comb.    Recal. Comb
-------------  --------  ---------  -------------
PreResNet-164  0.203799  0.109158       0.0910122
DenseNet-BC    0.147469  0.0894063      0.0740289


**OBSERVATIONS:**
* Combo Experiment is also able to provide similar results to Appendix D in paper

**Challenges Faced:**
* Output path needed to be changed.
* Code contained run functions for imagenet dataset. So, that needed to be commented.
* Directories mentioned were incorrect to read input data, so that also needed to be changed in code.
* Additional libariries were needed to be installed to run the code.

**Hyper-parameters**
* Output path parameter is changed
* cifar10h run function parameter was not called, so the cifar10h run function was called in the code.
* confusion matrix parameters for the human
* calibration parameters for the classifier
* seed parameter can be changed.




**BONUS**

**Insights from the paper:**


*   Author says that reason for hybrid human-machine approaches are for having trust between human and machine, and also because then one can take over if other lacks knowledge or expertise.
*   Human and machine don't make the same type of errors in image classification.
* Human labels and probabilistic predictions are independent given y.
* Two parameters are used : confusion matrix parameters for the human and calibration parameters for the classifier.
* 3 type of combinations are used : L+L, P+L, P+L-EM
* P+L performs the best.
* confidence ratios for machine and human labelers are also defined in the theoretical part of the paper.
* **Limitation:**
  * Experiment only performed on image classification. So, no guarantee of similar results on other kind of problems.



###Q.2. Model Multiple Humans:

In [ ]:
!python '/content/drive/MyDrive/conf_matrix_and_calibration-main/experiments/calibration_experiment2.py'

Models:   0% 0/4 [00:00<?, ?it/s]
Runs:   0% 0/25 [00:00<?, ?it/s]
Runs:   4% 1/25 [00:03<01:34,  3.95s/it]
Runs:   8% 2/25 [00:07<01:30,  3.94s/it]
Runs:  12% 3/25 [00:11<01:27,  3.98s/it]
Runs:  16% 4/25 [00:15<01:23,  3.98s/it]
Runs:  20% 5/25 [00:19<01:19,  3.97s/it]
Runs:  24% 6/25 [00:23<01:14,  3.93s/it]
Runs:  28% 7/25 [00:27<01:10,  3.94s/it]
Runs:  32% 8/25 [00:31<01:07,  3.96s/it]
Runs:  36% 9/25 [00:35<01:03,  3.96s/it]
Runs:  40% 10/25 [00:39<00:59,  3.95s/it]
Runs:  44% 11/25 [00:43<00:55,  3.96s/it]
Runs:  48% 12/25 [00:47<00:51,  3.93s/it]
Runs:  52% 13/25 [00:51<00:47,  3.94s/it]
Runs:  56% 14/25 [00:55<00:43,  3.96s/it]
Runs:  60% 15/25 [00:59<00:39,  3.97s/it]
Runs:  64% 16/25 [01:03<00:35,  3.96s/it]
Runs:  68% 17/25 [01:07<00:31,  3.97s/it]
Runs:  72% 18/25 [01:11<00:27,  3.96s/it]
Runs:  76% 19/25 [01:15<00:23,  3.95s/it]
Runs:  80% 20/25 [01:19<00:19,  3.94s/it]
Runs:  84% 21/25 [01:22<00:15,  3.90s/it]
Runs:  88% 22/25 [01:26<00:11,  3.88s/it]
Runs:  92% 23/25 [

**Approach for modelling more than one human on the CIFAR-10 dataset:**
* In the experiment file, call simulate_single_human function 3 times for 3 different humans by removing seed so that you don't get the same predicted label.
* Then you will have 3 different labels, compare all 3 labels index by index whichever prediction is in majority add it to the y_h label, and then call _run_experiment function. 
* If all 3 values are different then select one randomly.

**Output files shown below:**

In [ ]:
uploaded = files.upload()

Saving densenet-bc-L190-k40_accuracy.csv to densenet-bc-L190-k40_accuracy (1).csv


In [ ]:
densenet_q2 = pd.read_csv(io.BytesIO(uploaded['densenet-bc-L190-k40_accuracy.csv']))
(densenet_q2)

human     model  comb none  comb confusion  comb temperature scaling  \
0   0.967333  0.963667   0.979333        0.966667                  0.979333   
1   0.968000  0.966333   0.979667        0.970000                  0.979333   
2   0.966000  0.970333   0.982000        0.969000                  0.981667   
3   0.968000  0.972333   0.984333        0.973000                  0.982000   
4   0.965333  0.964333   0.980000        0.966333                  0.978000   
5   0.973667  0.968667   0.982000        0.972667                  0.984000   
6   0.972667  0.969333   0.983333        0.971667                  0.983667   
7   0.970667  0.971000   0.981000        0.975667                  0.982333   
8   0.969000  0.965333   0.977333        0.968667                  0.979333   
9   0.969667  0.966333   0.977333        0.969333                  0.978667   
10  0.970667  0.967333   0.982000        0.970333                  0.983333   
11  0.971000  0.972333   0.980333        0.971667                  0.982333   
12  0.970333  0.964000   0.978333        0.968333                  0.982333   
13  0.969333  0.965667   0.983000        0.970667                  0.981667   
14  0.971667  0.965000   0.982000        0.973333                  0.983667   
15  0.973000  0.968667   0.981667        0.971333                  0.982667   
16  0.967333  0.968333   0.981000        0.974000                  0.980333   
17  0.974333  0.963000   0.980000        0.972667                  0.980333   
18  0.972667  0.967000   0.981667        0.972333                  0.984333   
19  0.963667  0.969000   0.980333        0.968333                  0.980333   
20  0.965000  0.968333   0.980667        0.969000                  0.981000   
21  0.969667  0.973333   0.982333        0.972333                  0.982000   
22  0.968000  0.967667   0.982333        0.969667                  0.982000   
23  0.968667  0.962000   0.978000        0.967000                  0.978000   
24  0.968667  0.967667   0.980333        0.970333                  0.982333   

    comb ensemble temperature scaling  comb imax binning  
0                            0.978000           0.980667  
1                            0.978333           0.978667  
2                            0.980667           0.982000  
3                            0.979333           0.983333  
4                            0.974667           0.977667  
5                            0.982667           0.985667  
6                            0.982333           0.984667  
7                            0.982667           0.982333  
8                            0.978333           0.979667  
9                            0.977667           0.979333  
10                           0.982333           0.983667  
11                           0.980667           0.982333  
12                           0.981667           0.982000  
13                           0.981333           0.982000  
14                           0.984000           0.985333  
15                           0.981333           0.983333  
16                           0.978000           0.980333  
17                           0.980000           0.980000  
18                           0.982000           0.984000  
19                           0.977000           0.980000  
20                           0.979333           0.982000  
21                           0.980333           0.981667  
22                           0.980667           0.983000  
23                           0.977667           0.979333  
24                           0.981333           0.983333

In [ ]:
uploaded = files.upload()

Saving densenet-bc-L190-k40_calibration.csv to densenet-bc-L190-k40_calibration.csv


In [ ]:
densenet_q2_calib = pd.read_csv(io.BytesIO(uploaded['densenet-bc-L190-k40_calibration.csv']))
(densenet_q2_calib)

calibration method       metric     model      comb
0                 none      ECE (W)  0.028971  0.017101
1                 none      ECE (M)  0.026739  0.016482
2                 none   cwECE (WT)  0.040232  0.022960
3                 none   cwECE (MT)  0.024069  0.015553
4                 none  cwECE (WNT)  0.006635  0.003897
..                 ...          ...       ...       ...
870       imax binning   cwECE (WT)  0.025019  0.019520
871       imax binning   cwECE (MT)  0.014573  0.008549
872       imax binning  cwECE (WNT)  0.003570  0.002843
873       imax binning  cwECE (MNT)  0.001939  0.001214
874       imax binning          NLL  0.117004  0.503471

[875 rows x 4 columns]

In [ ]:
uploaded = files.upload()

Saving preresnet-110_accuracy.csv to preresnet-110_accuracy.csv


In [ ]:
preresnet_q2 = pd.read_csv(io.BytesIO(uploaded['preresnet-110_accuracy.csv']))
(preresnet_q2)

human     model  comb none  comb confusion  comb temperature scaling  \
0   0.970000  0.948667   0.974000        0.968333                  0.981000   
1   0.969667  0.950667   0.973000        0.967667                  0.982000   
2   0.969333  0.949000   0.976667        0.965667                  0.981000   
3   0.976000  0.953000   0.976667        0.972667                  0.983667   
4   0.968667  0.944000   0.969667        0.967667                  0.978000   
5   0.974000  0.952667   0.977333        0.972333                  0.982333   
6   0.971667  0.956000   0.976333        0.968667                  0.982667   
7   0.976333  0.950667   0.977000        0.972000                  0.987000   
8   0.973333  0.951000   0.976333        0.971000                  0.979667   
9   0.973000  0.950000   0.973667        0.969667                  0.981000   
10  0.966333  0.952667   0.975333        0.965333                  0.977333   
11  0.973000  0.958333   0.979000        0.973000                  0.984333   
12  0.971333  0.949333   0.971667        0.967333                  0.980333   
13  0.972000  0.948333   0.975333        0.968667                  0.981667   
14  0.975333  0.955333   0.976000        0.972000                  0.982667   
15  0.975333  0.952667   0.976667        0.969333                  0.983667   
16  0.972667  0.946667   0.975000        0.972333                  0.982000   
17  0.975000  0.949667   0.972667        0.970333                  0.980333   
18  0.971000  0.947000   0.973000        0.966000                  0.982667   
19  0.969333  0.944333   0.971333        0.963667                  0.979667   
20  0.969333  0.947667   0.972667        0.965667                  0.978000   
21  0.970667  0.952333   0.978000        0.969000                  0.982333   
22  0.970667  0.946000   0.973000        0.967000                  0.982667   
23  0.969000  0.944667   0.974333        0.963000                  0.982333   
24  0.975000  0.950000   0.975333        0.969000                  0.984667   

    comb ensemble temperature scaling  comb imax binning  
0                            0.981333           0.980333  
1                            0.982333           0.981000  
2                            0.981000           0.981333  
3                            0.983667           0.983000  
4                            0.978333           0.978000  
5                            0.981667           0.982667  
6                            0.983000           0.982667  
7                            0.986333           0.986000  
8                            0.980333           0.980333  
9                            0.982333           0.980333  
10                           0.977667           0.977667  
11                           0.984333           0.985667  
12                           0.980333           0.979667  
13                           0.981667           0.981000  
14                           0.982667           0.980333  
15                           0.983333           0.982667  
16                           0.982000           0.981333  
17                           0.980333           0.979333  
18                           0.982333           0.982333  
19                           0.979000           0.979667  
20                           0.978333           0.977000  
21                           0.982333           0.982333  
22                           0.983667           0.982667  
23                           0.982667           0.982667  
24                           0.984000           0.985000

In [ ]:
uploaded = files.upload()

Saving preresnet-110_calibration.csv to preresnet-110_calibration.csv


In [ ]:
preresnet_q2_calib = pd.read_csv(io.BytesIO(uploaded['preresnet-110_calibration.csv']))
(preresnet_q2_calib)

calibration method       metric     model      comb
0                 none      ECE (W)  0.030581  0.020960
1                 none      ECE (M)  0.030187  0.018874
2                 none   cwECE (WT)  0.047635  0.028853
3                 none   cwECE (MT)  0.026016  0.015371
4                 none  cwECE (WNT)  0.007819  0.004520
..                 ...          ...       ...       ...
870       imax binning   cwECE (WT)  0.030479  0.016217
871       imax binning   cwECE (MT)  0.016856  0.008955
872       imax binning  cwECE (WNT)  0.004583  0.002633
873       imax binning  cwECE (MNT)  0.002756  0.001177
874       imax binning          NLL  0.163614  0.438614

[875 rows x 4 columns]

In [ ]:
#Calculate error for multiple human model
import numpy as np
model_q2 = ['PreResNet-164', 'DenseNet-BC']
model_res = [(1 - np.mean(preresnet_q2['model']))*100, (1 - np.mean(densenet_q2['model']))*100]
print(model_res)
TS_calib = [(1 - np.mean(preresnet_q2['comb temperature scaling']))*100, (1 - np.mean(densenet_q2['comb temperature scaling']))*100]
print(TS_calib)
ETS_calib = [(1 - np.mean(preresnet_q2['comb ensemble temperature scaling']))*100, (1 - np.mean(densenet_q2['comb ensemble temperature scaling']))*100]
print(ETS_calib)
IMax_calib = [(1 - np.mean(preresnet_q2['comb imax binning']))*100, (1 - np.mean(densenet_q2['comb imax binning']))*100]
print(IMax_calib)

[4.997333333333332, 3.2519999999999993]
[1.8279999999999852, 1.859999999999995]
[1.8200000000000216, 1.9906666666666517]
[1.8600000000000172, 1.8146666666666866]


In [ ]:
from tabulate import tabulate
print('Results for Multiple Human Models')
info = {'Model Name': model_q2, 'Model': model_res, 'TS': TS_calib, 'ETS': ETS_calib, 'IMax': IMax_calib}
print(tabulate(info, headers='keys'))

Results for Multiple Human Models
Model Name       Model     TS      ETS     IMax
-------------  -------  -----  -------  -------
PreResNet-164  4.99733  1.828  1.82     1.86
DenseNet-BC    3.252    1.86   1.99067  1.81467


**OBSERVATIONS:**
* The error rates have been reduced by taking multiple human models.
* Multiple humans model is giving better accuracy than a signle human model.

**use-cases or application with such an approach where there
are more than one human decision makers in a human-AI team:**
* In health care system where multiple humans decisions are required like medical image diagnosis.
* In autonomous driving
* In stock prediction where inputs from humans can also be taken.
* In providing feedback or customer service.
* In course selection for college.
* In activities selection for co-curricular activities.
* In decision making for chatbots to make chats more real or emotional.
* In designing of a product where opinions of multiple humans are considered

**BONUS**

**Provide a better mathematical model to incorporate multiple humans based on
the theory proposed in the paper.**



*   As opinions or predictions of all the humans should not be given equal importance, so assign some weight to each human. 
* Like for medical diagonsis problem, prediction of a senior Doctor should be given more weightage than a new doctor or a nurse.
* So like this, instead of giving equal importance to predictions of all humans, different weights can be assigned to different predcitions of different humans.





### Now consider another case where the final “calibrated probabilities” are produced using a neural network model. Let’s call this neural network model as Team model. Here, the output of the Team model will have same units as the output of the machine learning model however the input to the Team model will be the output of machine learning model and human predicted label (consider the base case of only one human).
###Train the Team model and regenerate the results as in the base-case (Q.1). Compare your resultswith the base-case and list your observations. Do you find the Team model a better approach to combine human and machine model outputs?

In [ ]:
import os
os.chdir('/content/drive/MyDrive/conf_matrix_and_calibration-main')


In [ ]:
from data_utils import *
from utils import *
from keras.models import Sequential
from keras.layers import Dense

In [ ]:
from combination_methods import *
from metrics import *
from sklearn.model_selection import train_test_split
from metrics import *
import csv
import pandas as pd


In [ ]:
#Function to train Model
def myTeamModel(X, y):
  mymodel = Sequential()
  mymodel.add(Dense(64, input_dim=X.shape[1], activation='relu'))
  mymodel.add(Dense(32, activation='relu'))
  mymodel.add(Dense(10, activation='softmax'))
  mymodel.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
  mymodel.fit(X, y, epochs=10, batch_size=1024)
  return mymodel

In [ ]:
calibration_methods = ['none', 'confusion', 'temperature scaling', 'ensemble temperature scaling', 'imax binning']
calibration_metrics = {'ECE (W)': lambda probs, y: get_ECE(probs, y, mode='width'),
                          'ECE (M)': lambda probs, y: get_ECE(probs, y, mode='mass'),
                          'cwECE (WT)': lambda probs, y: get_cw_ECE(probs, y, mode='width'),
                          'cwECE (MT)': lambda probs, y: get_cw_ECE(probs, y, mode='mass'),
                          'cwECE (WNT)': lambda probs, y: get_cw_ECE(probs, y, mode='width', threshold_mode=None),
                          'cwECE (MNT)': lambda probs, y: get_cw_ECE(probs, y, mode='mass', threshold_mode=None),
                          'NLL': get_NLL}

In [ ]:
def _run_experiment(y_h=None, model_probs=None, y_true=None, **kwargs):
    seed = kwargs.pop('seed', 0)
    n_runs = kwargs.pop('n_runs', 25)
    test_size = kwargs.pop('test_size', 0.3)
    calibration_methods = kwargs.pop('calibration_methods', ['none'])
    calibration_metrics = kwargs.pop('calibration_metrics', {'ECE': get_ECE})
    output_file_acc = kwargs.pop('output_file_acc', './acc.csv')
    output_file_calibration = kwargs.pop('output_file_calibration', './cal.csv')

    acc_data = []
    cal_data = []
    for i in tqdm(range(n_runs), leave=False, desc='Runs'):
        # Train/test split
        y_h_tr, y_h_te, model_probs_tr, model_probs_te, y_true_tr, y_true_te = train_test_split(
            y_h, model_probs, y_true, test_size=test_size, random_state=i * seed)

        acc_h = get_acc(y_h_te, y_true_te)
        acc_m = get_acc(np.argmax(model_probs_te, axis=1), y_true_te)

        _acc_data = [acc_h, acc_m]
        _cal_data = []
        for calibration_method in calibration_methods:
            if calibration_method == 'confusion':
                combiner = DoubleConfusionCombiner()
                combiner.fit(model_probs_tr, y_h_tr, y_true_tr)
            else:
                combiner = OracleCombiner(calibration_method=calibration_method)
                combiner.fit(model_probs_tr, y_h_tr, y_true_tr)

            y_comb_te = combiner.combine(model_probs_te, y_h_te)
            acc_comb = get_acc(y_comb_te, y_true_te)
            _acc_data.append(acc_comb)

            model_probs_calibrated_te = combiner.calibrate(model_probs_te)
            y_comb_prob_te = combiner.combine_proba(model_probs_te, y_h_te)
            for metric, fxn in calibration_metrics.items():
                cal_m = fxn(model_probs_calibrated_te, y_true_te)
                cal_comb = fxn(y_comb_prob_te, y_true_te)
                _cal_data.append([calibration_method, metric, cal_m, cal_comb])

        acc_data += [_acc_data]
        cal_data += _cal_data

    # Save data to CSV
    header_acc = ['human', 'model'] + [f'comb {cal_m}' for cal_m in calibration_methods]
    with open(output_file_acc, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(header_acc)
        writer.writerows(acc_data)
    header_cal = ['calibration method', 'metric', 'model', 'comb']
    with open(output_file_calibration, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(header_cal)
        writer.writerows(cal_data)


In [ ]:
from tensorflow.keras.utils import to_categorical
from tqdm import tqdm
out_fpath = 'output-team'
experiment_args = {'n_runs': 25,
            'test_size': 0.3,
            'calibration_methods': calibration_methods,
            'calibration_metrics': calibration_metrics,
            'seed': 0
            }
model_names = ['r_low_acc', 'resnet-110', 'preresnet-110', 'densenet-bc-L190-k40']
for model_name in tqdm(model_names, desc='Models', leave=True):
    # Specify output files
    output_file_acc = out_fpath + f'{model_name}_accuracy.csv'
    output_file_calibration = out_fpath + f'{model_name}_calibration.csv'
    assert not os.path.exists(output_file_acc), 'Output filepath already exists'
    assert not os.path.exists(output_file_calibration), 'Output filepath already exists'
    experiment_args['output_file_acc'] = output_file_acc
    experiment_args['output_file_calibration'] = output_file_calibration

    # Load data
    human_counts, model_probs, y_true = load_CIFAR10H(model_name)

    X = np.hstack([model_probs, human_counts])
    y = to_categorical(y_true)
    y_h = simulate_single_human(human_counts, seed=0)
    train_model = myTeamModel(X, y)
    train_prob = train_model.predict(X)
    _run_experiment(y_h=y_h, model_probs=train_prob, y_true=y_true, **experiment_args)

Models:   0%|          | 0/4 [00:00<?, ?it/s]

Epoch 1/10
10/10 [==============================] - 3s 5ms/step - loss: 9.8853 - accuracy: 0.1159
Epoch 2/10
10/10 [==============================] - 0s 4ms/step - loss: 3.6175 - accuracy: 0.4044
Epoch 3/10
10/10 [==============================] - 0s 4ms/step - loss: 1.0960 - accuracy: 0.7367
Epoch 4/10
10/10 [==============================] - 0s 4ms/step - loss: 0.3266 - accuracy: 0.8960
Epoch 5/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0917 - accuracy: 0.9890
Epoch 6/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0525 - accuracy: 0.9905
Epoch 7/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0447 - accuracy: 0.9913
Epoch 8/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0404 - accuracy: 0.9918
Epoch 9/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0378 - accuracy: 0.9925
Epoch 10/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0357 - accuracy: 0.9927



Models:  25%|██▌       | 1/4 [02:06<06:18, 126.14s/it]

Epoch 1/10
10/10 [==============================] - 1s 4ms/step - loss: 2.7719 - accuracy: 0.5043
Epoch 2/10
10/10 [==============================] - 0s 4ms/step - loss: 0.6800 - accuracy: 0.8857
Epoch 3/10
10/10 [==============================] - 0s 4ms/step - loss: 0.1154 - accuracy: 0.9738
Epoch 4/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0507 - accuracy: 0.9899
Epoch 5/10
10/10 [==============================] - 0s 5ms/step - loss: 0.0406 - accuracy: 0.9913
Epoch 6/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0361 - accuracy: 0.9919
Epoch 7/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0335 - accuracy: 0.9925
Epoch 8/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0321 - accuracy: 0.9927
Epoch 9/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0309 - accuracy: 0.9926
Epoch 10/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0297 - accuracy: 0.9928



Models:  50%|█████     | 2/4 [04:03<04:01, 120.77s/it]

Epoch 1/10
10/10 [==============================] - 1s 4ms/step - loss: 5.2853 - accuracy: 0.1857
Epoch 2/10
10/10 [==============================] - 0s 4ms/step - loss: 1.0719 - accuracy: 0.6672
Epoch 3/10
10/10 [==============================] - 0s 4ms/step - loss: 0.1106 - accuracy: 0.9876
Epoch 4/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0481 - accuracy: 0.9908
Epoch 5/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0390 - accuracy: 0.9918
Epoch 6/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0355 - accuracy: 0.9926
Epoch 7/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0337 - accuracy: 0.9930
Epoch 8/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0325 - accuracy: 0.9933
Epoch 9/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0315 - accuracy: 0.9934
Epoch 10/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0306 - accuracy: 0.9934



Models:  75%|███████▌  | 3/4 [05:53<01:55, 115.78s/it]

Epoch 1/10
10/10 [==============================] - 1s 4ms/step - loss: 7.3662 - accuracy: 0.1498
Epoch 2/10
10/10 [==============================] - 0s 4ms/step - loss: 2.6659 - accuracy: 0.5325
Epoch 3/10
10/10 [==============================] - 0s 4ms/step - loss: 0.5041 - accuracy: 0.8541
Epoch 4/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0910 - accuracy: 0.9883
Epoch 5/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0561 - accuracy: 0.9898
Epoch 6/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0448 - accuracy: 0.9915
Epoch 7/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0399 - accuracy: 0.9926
Epoch 8/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0370 - accuracy: 0.9930
Epoch 9/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0351 - accuracy: 0.9936
Epoch 10/10
10/10 [==============================] - 0s 4ms/step - loss: 0.0337 - accuracy: 0.9936



Models: 100%|██████████| 4/4 [07:42<00:00, 115.70s/it]


**Output files shown below:**

In [ ]:
uploaded = files.upload()

Saving output-teamdensenet-bc-L190-k40_accuracy.csv to output-teamdensenet-bc-L190-k40_accuracy (2).csv


In [ ]:
import pandas as pd
import io
teamdensenet_acc = pd.read_csv(io.BytesIO(uploaded['output-teamdensenet-bc-L190-k40_accuracy.csv']))
(teamdensenet_acc)

human  model  comb none  comb confusion  comb temperature scaling  \
0   0.955333  0.991      0.984        0.988667                  0.985333   
1   0.955333  0.991      0.984        0.988667                  0.985333   
2   0.955333  0.991      0.984        0.988667                  0.985333   
3   0.955333  0.991      0.984        0.988667                  0.985333   
4   0.955333  0.991      0.984        0.988667                  0.985333   
5   0.955333  0.991      0.984        0.988667                  0.985333   
6   0.955333  0.991      0.984        0.988667                  0.985333   
7   0.955333  0.991      0.984        0.988667                  0.985333   
8   0.955333  0.991      0.984        0.988667                  0.985333   
9   0.955333  0.991      0.984        0.988667                  0.985333   
10  0.955333  0.991      0.984        0.988667                  0.985333   
11  0.955333  0.991      0.984        0.988667                  0.985333   
12  0.955333  0.991      0.984        0.988667                  0.985333   
13  0.955333  0.991      0.984        0.988667                  0.985333   
14  0.955333  0.991      0.984        0.988667                  0.985333   
15  0.955333  0.991      0.984        0.988667                  0.985333   
16  0.955333  0.991      0.984        0.988667                  0.985333   
17  0.955333  0.991      0.984        0.988667                  0.985333   
18  0.955333  0.991      0.984        0.988667                  0.985333   
19  0.955333  0.991      0.984        0.988667                  0.985333   
20  0.955333  0.991      0.984        0.988667                  0.985333   
21  0.955333  0.991      0.984        0.988667                  0.985333   
22  0.955333  0.991      0.984        0.988667                  0.985333   
23  0.955333  0.991      0.984        0.988667                  0.985333   
24  0.955333  0.991      0.984        0.988667                  0.985333   

    comb ensemble temperature scaling  comb imax binning  
0                            0.985333           0.985667  
1                            0.985333           0.985667  
2                            0.985333           0.985667  
3                            0.985333           0.985667  
4                            0.985333           0.985667  
5                            0.985333           0.985667  
6                            0.985333           0.985667  
7                            0.985333           0.985667  
8                            0.985333           0.985667  
9                            0.985333           0.985667  
10                           0.985333           0.985667  
11                           0.985333           0.985667  
12                           0.985333           0.985667  
13                           0.985333           0.985667  
14                           0.985333           0.985667  
15                           0.985333           0.985667  
16                           0.985333           0.985667  
17                           0.985333           0.985667  
18                           0.985333           0.985667  
19                           0.985333           0.985667  
20                           0.985333           0.985667  
21                           0.985333           0.985667  
22                           0.985333           0.985667  
23                           0.985333           0.985667  
24                           0.985333           0.985667

In [ ]:
uploaded = files.upload()

Saving output-teamdensenet-bc-L190-k40_calibration.csv to output-teamdensenet-bc-L190-k40_calibration (1).csv


In [ ]:
teamdensenet_calib = pd.read_csv(io.BytesIO(uploaded['output-teamdensenet-bc-L190-k40_calibration.csv']))
(teamdensenet_calib)

calibration method       metric     model      comb
0                 none      ECE (W)  0.006777  0.011339
1                 none      ECE (M)  0.005656  0.010468
2                 none   cwECE (WT)  0.015063  0.018604
3                 none   cwECE (MT)  0.007654  0.010728
4                 none  cwECE (WNT)  0.002198  0.003029
..                 ...          ...       ...       ...
870       imax binning   cwECE (WT)  0.009739  0.017595
871       imax binning   cwECE (MT)  0.005718  0.009309
872       imax binning  cwECE (WNT)  0.001471  0.002764
873       imax binning  cwECE (MNT)  0.000897  0.001279
874       imax binning          NLL  0.042004  0.478171

[875 rows x 4 columns]

In [ ]:
uploaded = files.upload()

Saving output-teampreresnet-110_accuracy.csv to output-teampreresnet-110_accuracy (1).csv


In [ ]:
teampreresnet_acc = pd.read_csv(io.BytesIO(uploaded['output-teampreresnet-110_accuracy.csv']))
(teampreresnet_acc)

human     model  comb none  comb confusion  comb temperature scaling  \
0   0.955333  0.992667   0.985333        0.990667                     0.986   
1   0.955333  0.992667   0.985333        0.990667                     0.986   
2   0.955333  0.992667   0.985333        0.990667                     0.986   
3   0.955333  0.992667   0.985333        0.990667                     0.986   
4   0.955333  0.992667   0.985333        0.990667                     0.986   
5   0.955333  0.992667   0.985333        0.990667                     0.986   
6   0.955333  0.992667   0.985333        0.990667                     0.986   
7   0.955333  0.992667   0.985333        0.990667                     0.986   
8   0.955333  0.992667   0.985333        0.990667                     0.986   
9   0.955333  0.992667   0.985333        0.990667                     0.986   
10  0.955333  0.992667   0.985333        0.990667                     0.986   
11  0.955333  0.992667   0.985333        0.990667                     0.986   
12  0.955333  0.992667   0.985333        0.990667                     0.986   
13  0.955333  0.992667   0.985333        0.990667                     0.986   
14  0.955333  0.992667   0.985333        0.990667                     0.986   
15  0.955333  0.992667   0.985333        0.990667                     0.986   
16  0.955333  0.992667   0.985333        0.990667                     0.986   
17  0.955333  0.992667   0.985333        0.990667                     0.986   
18  0.955333  0.992667   0.985333        0.990667                     0.986   
19  0.955333  0.992667   0.985333        0.990667                     0.986   
20  0.955333  0.992667   0.985333        0.990667                     0.986   
21  0.955333  0.992667   0.985333        0.990667                     0.986   
22  0.955333  0.992667   0.985333        0.990667                     0.986   
23  0.955333  0.992667   0.985333        0.990667                     0.986   
24  0.955333  0.992667   0.985333        0.990667                     0.986   

    comb ensemble temperature scaling  comb imax binning  
0                               0.986              0.986  
1                               0.986              0.986  
2                               0.986              0.986  
3                               0.986              0.986  
4                               0.986              0.986  
5                               0.986              0.986  
6                               0.986              0.986  
7                               0.986              0.986  
8                               0.986              0.986  
9                               0.986              0.986  
10                              0.986              0.986  
11                              0.986              0.986  
12                              0.986              0.986  
13                              0.986              0.986  
14                              0.986              0.986  
15                              0.986              0.986  
16                              0.986              0.986  
17                              0.986              0.986  
18                              0.986              0.986  
19                              0.986              0.986  
20                              0.986              0.986  
21                              0.986              0.986  
22                              0.986              0.986  
23                              0.986              0.986  
24                              0.986              0.986

In [ ]:
uploaded = files.upload()

Saving output-teampreresnet-110_calibration.csv to output-teampreresnet-110_calibration (1).csv


In [ ]:
teampreresnet_calib = pd.read_csv(io.BytesIO(uploaded['output-teampreresnet-110_calibration.csv']))
(teampreresnet_calib)

calibration method       metric     model      comb
0                 none      ECE (W)  0.005050  0.012094
1                 none      ECE (M)  0.003994  0.010234
2                 none   cwECE (WT)  0.013198  0.018512
3                 none   cwECE (MT)  0.005640  0.008886
4                 none  cwECE (WNT)  0.002078  0.002861
..                 ...          ...       ...       ...
870       imax binning   cwECE (WT)  0.008762  0.015877
871       imax binning   cwECE (MT)  0.003556  0.009021
872       imax binning  cwECE (WNT)  0.001509  0.002629
873       imax binning  cwECE (MNT)  0.000732  0.001021
874       imax binning          NLL  0.046353  0.445510

[875 rows x 4 columns]

In [ ]:
#Errors
model_q3 = ['PreResNet-164', 'DenseNet-BC']
model_res = [(1 - np.mean(teampreresnet_acc['model']))*100, (1 - np.mean(teamdensenet_acc['model']))*100]
print(model_res)
TS_calib = [(1 - np.mean(teampreresnet_acc['comb temperature scaling']))*100, (1 - np.mean(teamdensenet_acc['comb temperature scaling']))*100]
print(TS_calib)
ETS_calib = [(1 - np.mean(teampreresnet_acc['comb ensemble temperature scaling']))*100, (1 - np.mean(teamdensenet_acc['comb ensemble temperature scaling']))*100]
print(ETS_calib)
IMax_calib = [(1 - np.mean(teampreresnet_acc['comb imax binning']))*100, (1 - np.mean(teamdensenet_acc['comb imax binning']))*100]
print(IMax_calib)

[0.7333333333332748, 0.900000000000023]
[1.399999999999968, 1.4666666666666828]
[1.399999999999968, 1.4666666666666828]
[1.399999999999968, 1.4333333333333198]


In [ ]:
print('Results for TeamModel')
info = {'Model Name': model_calib, 'Model': model_res, 'TS': TS_calib, 'ETS': ETS_calib, 'IMax': IMax_calib}
print(tabulate(info, headers='keys'))

Results for TeamModel
Model Name        Model       TS      ETS     IMax
-------------  --------  -------  -------  -------
PreResNet-164  0.733333  1.4      1.4      1.4
DenseNet-BC    0.9       1.46667  1.46667  1.43333


**OBSERVATIONS:**
* The error have reduced significantly than our base model.
* We can clearly see Team model gives lesser than our previous base-case.
* So, Team model gives a better approach to combine human and machine model outputs.